
# __Física Computacional:__ _Tarea 1_

### Joshua Briones Andrade
#### Fecha de entrega: _9 de Octubre_
----

### Ejercicio 1

Escribir un programa que almacene en una **lista** los números del $1$ al $10$ y los muestre en pantalla en orden inverso separados por comas.

In [2]:
# Primero creamos una lista vacia llamada "list1"
list1 = []

# Ahora usando un ciclo "for", agregamos los valores de 1 a 10 con el método "append"
for i in range(10):   
    list1.append(i+1)

# También se puede hacer en una linea usando "list comprehensions"
list2 = [k + 1 for k in range(10)]

# Aprovechando un truco de "slicing" sobre las listas, cambiamos el orden de "list1" y lo imprimimos con "print()"
print(list1[::-1])

[10, 9, 8, 7, 6, 5, 4, 3, 2, 1]


### Ejercicio 2
Escribir un programa que almacene el abecedario en una lista, elimine de la lista las letras que ocupen posiciones múltiplos de 3, y muestre por pantalla la lista resultante.

In [7]:
# CORREGIDO

abc = ["a","b","c","d","e","f","g","h","i","j","k","l","m","n","o","p","q","r","s","t","u","v","w","x","y","z"]

def muestra_lista(abc):
    temp = []
    for i in range(len(abc)):
        if i % 3 == 0:
            temp.append(i)
    for i in temp[::-1]:
        abc.pop(i)
    print("La lista resutante es:", abc)
    
muestra_lista(abc)

La lista resutante es: ['b', 'c', 'e', 'f', 'h', 'i', 'k', 'l', 'n', 'o', 'q', 'r', 't', 'u', 'w', 'x', 'z']


### Ejercicio 3
Escribir un programa que pida al usuario una palabra y muestre por pantalla si es un
palíndromo

In [1]:
# CORREGIDO

def palindromo():
    frase = input("Dame una frase, camarada: ")
    string = frase.lower().replace(" ","").replace(",","")
    n = len(string); s = True
    
    for i in range(n):
        if string[i] != string[(n-1)-i]:
            s = False

    if s == False:
        print(f"La frase \"{frase}\" NO es un palíndromo :c")
    else:
        print(f"¡Wow!, la frase \"{frase}\" SÍ es un palíndromo c:")

In [2]:
palindromo()
palindromo()

Dame una frase, camarada: Oso oso
¡Wow!, la frase "Oso oso" SÍ es un palíndromo c:
Dame una frase, camarada: mic cim
¡Wow!, la frase "mic cim" SÍ es un palíndromo c:


### Ejercicio 4
Pide números y mételos en una lista, cuando el usuario meta un 0 ya dejaremos de
insertar. Por último, muestra los números ordenados de menor a mayor.

In [14]:
def pedir_numeros(hist = None):
    if hist == None:
        hist = []
    numero = input("Dame un número. Si es un \"0\", me detengo: ")
    if float(numero) == 0: 
        hist.sort(reverse=False)
        print("\nHistorial de números: ", hist)
    else:
        hist.append(float(numero))
        pedir_numeros(hist=hist)
        
pedir_numeros()

Dame un número. Si es un "0", me detengo: 1
Dame un número. Si es un "0", me detengo: 42
Dame un número. Si es un "0", me detengo: 27
Dame un número. Si es un "0", me detengo: 0

Historial de números:  [1.0, 27.0, 42.0]


### Ejercicio 5
Escribir un programa que pida al usuario una palabra y muestre por pantalla el número de veces que contiene cada vocal.

In [59]:
def cuenta_vocales():
    human_word = input("Deme una palabra, por favor humano: ").lower()
    human_dict = {}

    for i in range(len( human_word )):
        if human_word[i] in ["a","e","i","o","u"]:
            human_dict[human_word[i]]=human_dict.get(human_word[i],0)+1
        else: pass
    return human_dict

cuenta_vocales()

Deme una palabra, por favor humano: Murcielago


{'u': 1, 'i': 1, 'e': 1, 'a': 1, 'o': 1}

### Ejercicio 6
Pide una cadena por teclado, mete los caracteres en una lista sin espacios.

In [22]:
def string_to_list():
    cadena = input("Amo Bruce, permitame su cadena de texto por favor: \n")
    cadena = cadena.replace(" ","")
    listaa = [i for i in cadena]
    return listaa

string_to_list()

Amo Bruce, permitame su cadena de texto por favor: 
Claro que sí, Alfred


['C',
 'l',
 'a',
 'r',
 'o',
 'q',
 'u',
 'e',
 's',
 'í',
 ',',
 'A',
 'l',
 'f',
 'r',
 'e',
 'd']

### Ejercicio 7
Haz un programa que pida el nombre de una persona y lo muestre en pantalla repetido 1000 veces, pero dejando un espacio de separación entre cada copia del nombre.

In [23]:
def repetir_nombre():
    supername = input("Oye, ¿cómo se llama esa persona cuyo nombre se tiene que decir 1000 veces?: ")
    superlist = (supername + " ")*1000

    print()
    print(superlist)
    
repetir_nombre()

Oye, ¿cómo se llama esa persona cuyo nombre se tiene que decir 1000 veces?: Voldemort

Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Voldemort Vol

### Ejercicio 8
Escribir un programa que guarde en una variable el diccionario "Euro": 25.82, "Dolar":
24.113, "Yen": 0.2189, "Dolar Canadiense": 16.7842 y le pregunte al usuario por una divisa y muestre su valor de cambio o un mensaje de aviso si la divisa no está en el diccionario. 

Completa con todas las divisas que puedas.

In [61]:
divisas = {"Euro": 25.82, "Dolar": 24.113, "Yen": 0.2189, "Dolar Canadiense": 16.7842, "Bitcoin": 228167.75, "Quetzal": 2.78}

def preguntar_divisa():
    divisa = input("¿En qué divisa esta interesado? Tenemos \"Euro\", \"Dolar\", \"Yen\", \"Dolar Canadiense\", \"Bitcoin\" y \"Quetzal\": \n")
    valor = divisas.get(divisa, 0)
    if valor == 0:
        print("Lo sentimos, no tenemos esa divisa en nuestro diccionario")
    else:
        print(f"El valor de cambio del {divisa}, es de {valor} en la moneda local.\n")      

preguntar_divisa()
preguntar_divisa()

¿En qué divisa esta interesado? Tenemos "Euro", "Dolar", "Yen", "Dolar Canadiense", "Bitcoin" y "Quetzal": 
Bitcoin
El valor de cambio del Bitcoin, es de 228167.75 en la moneda local.

¿En qué divisa esta interesado? Tenemos "Euro", "Dolar", "Yen", "Dolar Canadiense", "Bitcoin" y "Quetzal": 
Chelín
Lo sentimos, no tenemos esa divisa en nuestro diccionario


### Ejercicio 9
Escribir un programa que almacene el diccionario con las calificaciones de las asignaturas de un semestre "Algebra": 10, "Física": 9, "Calculo": 8, "Computacion": 10, "Geometria": 9, y después muestre por pantalla las calificaciones de cada asignatura en el formato asignatura tiene calificacion , donde asignatura es cada una de las asignaturas del curso, y calificacion son sus calificaciones. 

Al final debe mostrar también el número promedio del curso.

In [26]:
calif = {"Algebra": 10, "Física": 9, "Calculo": 8, "Computacion": 10, "Geometria": 9}

def leer_calificaciones(calif):
    s = 0
    for key in calif.items():
        print(f"En \"{key[0]}\" tiene de calificación: {key[1]}")
        s += key[1]
    print(f"\nEl promedio de calificaciones es {s/5}")
    
leer_calificaciones(calif)

En "Algebra" tiene de calificación: 10
En "Física" tiene de calificación: 9
En "Calculo" tiene de calificación: 8
En "Computacion" tiene de calificación: 10
En "Geometria" tiene de calificación: 9

El promedio de calificaciones es 9.2


### Ejercicio 10
Un satélite se encuentra en una órbita circular alrededor de la Tierra, tal que completa un círculo en $T$ segundos, $T$ es el periodo de la órbita:

1. Muestra que la altitud $h$ sobre la superficie de la Tierra será:


$$h=\left(\frac{GMT^2}{4\pi^2}\right)^{1/3}-R$$


donde $G$ es la constante de gravitación universal con valor $G = 6.67\times 10^{−11}\ m^3kg^{-1}s^{-2}$, $M$ es la masa de la Tierra $M = 5.97 \times 10^{24}\ kg$ y $R$ es el radio de la Tierra $R = 6371\ km$.

2. Escribe un programa que le pregunte al usuario el periodo que desea tenga su satélite girando alrededor de la Tierra y el programa le regrese la altitud (en metros) que deberá tener.

3. ¿Qué altitud deberá tener el satélite su queremos que orbite la Tierra con un periodo de un día (órbita geosíncrona), 90 minutos, 45 minutos? ¿Qué observas?

4. En la práctica, una órbita geosíncrona es aquella cuya órbita tiene un periodo de un día sideral, que es de 23.93 horas, no de 24 horas. ¿Por qué es esto? ¿Cuál será la diferencia en altitud entre $T= 23.93$ horas y $T= 24$ horas.


Para este problema, primero nos enfocamos en la fuerza que experimenta el satélite (en el eje radial)

$$F_T=F_C$$

donde $F_G$ es la fuerza gravitatoria entre Tierra-satélite, y $F_C$ la fuerza centrifuga del satélite. Con esto dicho podemos hacer el siguiente despeje.

$$G\frac{Mm}{r^2}=m\frac{v^2}{r}$$

donde $M$ es la masa de la Tierra, $m$ la masa del satélite, $v$ la velocidad radial del satélite y $r$ la distancia del centro de la Tierra al satélite. Simplificamos esta expresión.

\begin{equation*}
G\frac{Mm}{r^2}=m\frac{v^2}{r}\quad\Rightarrow
G\frac{M}{r}=v^2
\end{equation*}


Y como la velocidad radial esta definida como $v=\omega\cdot r=\left(\frac{2\pi}{T}\right)\cdot r$, sustituyendo en la ecuación, despejamos $r$

\begin{equation*}
G\frac{M}{r}=\left(\frac{2\pi}{T}\cdot r\right)^2=\frac{4\pi^2}{T^2}\cdot r^2\quad\Rightarrow 
r^3=\frac{GMT^2}{4\pi^2}\quad\Rightarrow r=\left(\frac{GMT^2}{4\pi^2}\right)^{1/3}
\end{equation*}

Dado que $r$ es la distancia del centro de la Tierra al satélite, se deduce que $r=R+h$, donde $R$ es el radio de la Tierra y $h$ es la distancia de la Tierra al satélite.

Por lo tanto, al resolver para $h$, se tiene la siguiente expresión.

$$\boxed{h=\left(\frac{GMT^2}{4\pi^2}\right)^{1/3}-R}$$

In [28]:
""" 
10.2. Para este problema tenemos que agregar las constantes previamente para no crear variables
cada que llamemos a la función, entonces simplemente es cuestión de aplicar la formula anterior
"""

from math import pi
G = 6.67 * 10 **(-11) # m^{3}kg^{-1}s^{-2}
M = 5.97 * 10 **(24)  # kg
R = 6371 * 10 **(3)   # m 

def h(T):
    """
    T debe estar en segundos [s] y el resultado esta en metros [m]
    """
    frac = (G*M*T**2)/(4*pi**2)
    h = frac**(1/3) - R
    return h

h(25*60*60)  # Altura de un satélite con un periodo de 25 horas

37020881.55756288

In [46]:
"""
10.3. Calculamos la altura para una órbita de t1=90 minutos y t2=45 minutos.
"""

t1 = 90   # minutos
t2 = 45   # minutos

print(f"Órbita de {t1} minutos: {round(h(t1*60)/1000,2)} km de altura \nÓrbita de {t2} minutos: {round(h(t2*60)/1000,2)} km de altura")

Órbita de 90 minutos: 279.32 km de altura 
Órbita de 45 minutos: -2181.56 km de altura


Para el problema **10.3.** podemos ver que la órbita de **90 minutos** si es posible pero es a apenas **280 km** de altura y para la órbita de **45 minutos**, se tiene una **altura negativa** (-2182 km), por lo que no es posible realizar dicha órbita porque chocaría con la Tierra.

In [52]:
"""
10.4. Para la parte de código, calculamos simplemente la diferencia de alturas entre estos dos periodos.
"""

h1 = h(23.93*(60*60))
h2 = h(24*(60*60))

(f"La diferencia de alturas entre una órbita de 24 y 23.93 horas es {round((h2-h1)/100,2)} km")

'La diferencia de alturas entre una órbita de 24 y 23.93 horas es 821.48 km'

La razón del porqué las órbitas geoestacionarias son de $23.93$ horas es debido estan ligadas a los días siderales que no son 24 horas.

Según Wikipedia, un <a src="https://es.wikipedia.org/wiki/D%C3%ADa_sid%C3%A9reo">día sideral</a> es el lapso transcurrido entre dos culminaciones (o tránsitos) sucesivos del Primer punto de Aries. [...] El día sidéreo es unos 4 minutos más corto que el día solar medio.

#### Órbita geoestacionaria:
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/4/49/Geostat.gif/250px-Geostat.gif">

Son de gran importancia para telecomunicaciones ya que permanecen en un solo punto en el cielo respecto a los observadores en la Tierra por lo que están ligados necesariamente a un periodo de un día sideral.